core/src/main/scala/kafka/log/Log.scala

```diff
@@ -527,8 +527,10 @@ class Log(val dir: File,
     if(numToDelete > 0) {
       lock synchronized {
         // we must always have at least one segment, so if we are going to delete all the segments, create a new one first
-        if(segments.size == numToDelete)
+        if(segments.size == numToDelete) {
           roll()
+          updateLogEndOffset(nextOffsetMetadata.messageOffset)
```

Comment:
This makes sense to me. I think it is OK to possibly call  do updateLogEndOffset() twice in the Log.append(). I have updated patch as you suggested.